### PR TITLE
Chan 478 persistent banner

### PR DESCRIPTION
**Pull request by tomsoir**
https://reddit.atlassian.net/browse/CHAN-478
https://reddit.slack.com/files/schwers/F4ZFNS0N6/persistentupdates.js#

### CODE REVIEW

File: src/app/actions/xpromo.js [hunk 1]
```diff
@@ -214,4 +233,4 @@ const trackAppStoreVisit = (visitType, extraData) => trackXPromoEvent(XPROMO_APP
  */
 export const navigateToAppStore = url => {
   window.location = url;
-};
+};
```
Comment (schwers) [inline src/app/actions/xpromo.js]: :nitpick: needs a new line

File: src/app/components/DualPartInterstitial/Footer/index.jsx [hunk 2]
```diff
@@ -111,7 +122,7 @@ class DualPartInterstitialFooter extends React.Component {
 }
 
 const selector = createStructuredSelector({
-  subredditName: getSubreddit,
+  subredditName: state => getSubreddit(state),
```
Comment (schwers) [inline src/app/components/DualPartInterstitial/Footer/index.jsx]: :nitpick: this seems equivalent, why the change?

File: src/app/components/Place/index.jsx [hunk 3]
```diff
@@ -57,7 +57,7 @@ export default connect(selector, mapDispatchToProps)(
 
     render() {
       const isSubredditPage = includes(this.props.currentPageUrl, '/r/place');
-      const placeUrl = config.placeDomain + '/place?webview=true';
+      const placeUrl = `${config.placeDomain }/place?webview=true`;
```
Comment (schwers) [inline src/app/components/Place/index.jsx]: :nitpick: extra whitespace before the closing `}`

File: src/app/components/Place/index.jsx [hunk 4]
```diff
@@ -42,7 +42,7 @@ export default connect(selector, mapDispatchToProps)(
             JSON.stringify({
               name: 'PLACE_MESSAGE',
               payload: {
-                Authorization: 'bearer ' + this.props.token,
+                Authorization: `bearer ${ this.props.token}`,
```
Comment (schwers) [inline src/app/components/Place/index.jsx]: :nitpick: extra whitespace after the opening `{`

File: src/app/reducers/accounts.js [hunk 5]
```diff
@@ -13,6 +13,31 @@ export default function(state=DEFAULT, action={}) {
 
     case accountActions.RECEIVED_ACCOUNT: {
       const { accounts } = action.apiResponse;
+
+      Object.keys(accounts).forEach(accountName => {
```
Comment (schwers) [inline src/app/reducers/accounts.js]: ☑️ reminder, let's remove before deploying

File: src/lib/getSubredditFromState.js [hunk 6]
```diff
@@ -1,6 +1,8 @@
 export default function getSubreddit(state) {
   if (state.platform.currentPage.urlParams.subredditName) {
     return state.platform.currentPage.urlParams.subredditName;
+  } else if (state.platform.currentPage.urlParams.subredditName === undefined) {
```
Comment (schwers) [inline src/lib/getSubredditFromState.js]: What's the reasoning for this change? The logic after this return statement should still be run to handle comments pages that don't have a subreddit name in the url

File: src/lib/xpromoPersistState.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    return true;
+  }
+  // If more then HIDE time and the session is new 
+  // — change to show the banner
+  if ((Date.now() > (lsTime + config.hideTime)) && !dismissedState) {
+    setLocalStorage();
+    callback(status.SHOW__NEW_SESSION);
+    return true;
+  } 
+  // For other cases we dont need to show up the bunner
```
Comment (schwers) [inline src/lib/xpromoPersistState.js]: `bunner` -> `banner`

File: src/app/featureFlags.js [hunk 8]
```diff
@@ -256,6 +259,22 @@ const config = {
       ] },
     ],
   },
+  [VARIANT_XPROMO_PERSISTENT_IOS]: {
+    and: [
+      { allowedDevices: [IPHONE] },
```
Comment (schwers) [inline src/app/featureFlags.js]: Should there be checks for NSFW and limiting it to listing / index pages?

File: src/app/featureFlags.js [hunk 9]
```diff
@@ -256,6 +259,22 @@ const config = {
...
+      { or: [
```
Comment (schwers) [inline src/app/featureFlags.js]: 🔕 Why nest this in an `or`? Should there be a `url` flag here?

Comment (schwers) [inline src/app/featureFlags.js]: Do we still want to exclude the `modal_listing_click` cohort?

File: src/app/selectors/xpromo.js [hunk 10]
```diff
@@ -317,7 +352,7 @@ export function shouldShowXPromo(state) {
 }
 
 export function isPartOfXPromoExperiment(state) {
-  return shouldShowXPromo(state) && !!activeXPromoExperimentName(state);
+  return shouldShowXPromo(state) && anyFlagEnabled(state);
```
Comment (schwers) [inline src/app/selectors/xpromo.js]: `anyFlagEnabled` has no default value, it looks like `EXPERIMENT_FULL` should be passed here?

File: src/lib/xpromoPersistState.js [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    !isStatusNotExist(statusKey.SHOW__SAME_SESSION) ||
+    !isStatusNotExist(statusKey.SHOW__NEW_SESSION)
+  );
+};
+
+/*
+ * Utils for working with localStorage
+ */ 
+const isXpromoClosed = () => {
+  if (localStorageAvailable()) {
```
Comment (schwers) [inline src/lib/xpromoPersistState.js]: This seems like it should be passed `state` so it can check the experiment bucket. Having an item in local storage does not imply that the interstitial is closed, it might have been closed 3 weeks ago and now be on screen

File: src/app/components/DualPartInterstitial/Footer/index.jsx (hunk 2 above)
Comment (tomsoir) [inline src/app/components/DualPartInterstitial/Footer/index.jsx]: These changes occurred after rebasing. Cleaned up, thank you.

**Comment by tomsoir** [comment]
@schwers @prashtx 💇‍♂️

File: src/app/constants.js [hunk 12]
```diff
@@ -145,11 +145,16 @@ export const flags = {
   VARIANT_XPROMO_INTERSTITIAL_FREQUENCY_ANDROID: 'experimentXPromoInterstitialFrequencyAndroid',
   VARIANT_XPROMO_INTERSTITIAL_FREQUENCY_IOS_CONTROL: 'experimentXPromoInterstitialFrequencyIosControl',
   VARIANT_XPROMO_INTERSTITIAL_FREQUENCY_ANDROID_CONTROL: 'experimentXPromoInterstitialFrequencyAndroidControl',
+
+  // XPromo Persistent Banner
+  VARIANT_XPROMO_PERSISTENT_IOS: 'experimentXPromoPersistentIos',
+  VARIANT_XPROMO_PERSISTENT_ANDROID: 'experimentXPromoPersistentAndroid',
 };
 
 export const xpromoDisplayTheme = {
```
Comment (schwers) [inline src/app/constants.js]: :nitpick: We case constants as all upper case, could you make this be `XPROMO_DISPLAY_THEMES`?

File: src/app/components/DualPartInterstitial/Footer/index.jsx [hunk 13]
```diff
@@ -7,8 +7,8 @@ import url from 'url';
 
 import { redirect } from 'platform/actions';
 import * as xpromoActions from 'app/actions/xpromo';
-import { XPROMO_DISMISS } from 'lib/eventUtils';
 import getSubreddit from 'lib/getSubredditFromState';
+import { xpromoDisplayTheme as theme } from 'app/constants';
```
Comment (schwers) [inline src/app/components/DualPartInterstitial/Footer/index.jsx]: :nitpick: once the constant is renamed, could you also rename `theme` -> `THEMES`

File: src/app/featureFlags.js [hunk 14]
```diff
@@ -256,6 +259,30 @@ const config = {
...
+  [VARIANT_XPROMO_PERSISTENT_IOS]: {
+    and: [
+      { allowedDevices: [IPHONE] },
+      { allowedPages: ['index', 'listing', 'comments'] },
+      { allowNSFW: false },
+      { or: [
+        { url: 'xpromopersistent' },
+        { variant: 'mweb_xpromo_persistent_ios:treatment' },
+      ] },
+      { not: { peak: 'mweb_xpromo_modal_listing_click_ios' } },
```
Comment (schwers) [inline src/app/featureFlags.js]: `and` will use `Array.prototype.every` under-the-hood to evaluate its list of rules. Could you move this check to line 266 so it comes before the variant check? That way we don't trigger the bucketing event if this `peak` rule would prevent us from being in the experiment

File: src/app/featureFlags.js [hunk 15]
```diff
@@ -256,6 +259,30 @@ const config = {
...
+  [VARIANT_XPROMO_PERSISTENT_ANDROID]: {
+    and: [
+      { allowedDevices: [ANDROID] },
+      { allowedPages: ['index', 'listing', 'comments'] },
+      { allowNSFW: false },
+      { or: [
+        { url: 'xpromopersistent' },
+        { variant: 'mweb_xpromo_persistent_android:treatment' },
+      ] },
+      { not: { peak: 'mweb_xpromo_modal_listing_click_android' } },
```
Comment (schwers) [inline src/app/featureFlags.js]: Same idea here with rule placement, this should be before the `variant` rule, could you move this to 278/9?

File: src/app/selectors/xpromo.js [hunk 16]
```diff
@@ -301,25 +319,33 @@ export function getXPromoExperimentPayload(state) {
...
 }
 
+export function dismissedState(state) {
+  return state.xpromo.persistent.dismissed;
+}
+export function isXPromoPersistentActive(state) {
+  return state.xpromo.persistent.active;
+}
+export function isXPromoPersistent(state) {
+  if (isInterstitialDimissed(state)) {
```
Comment (schwers) [inline src/app/selectors/xpromo.js]: The return type of this is either `boolean` or `undefined`. For this to play nice with `redux` and `reselect`, it would be better to always return the same type of value. ie. `undefined` is equivalent to returning `false`, so it would be good to just always return a `boolean`.

How about making the function be 
```js
return isInterstitialDimissed(state) && anyFlagEnabled(state, XPROMO_PERSISTENT_FLAGS);
```

File: src/lib/timeChecker.js [hunk 17]
```diff
@@ -0,0 +1,31 @@
+export default class TimeChecker {
```
Comment (schwers) [inline src/lib/timeChecker.js]: 🔕  The existence of this class and the `displayPersistBannerByTimer` function is a strong example of why having `redux-observeable or `redux-sagas` in the code base would be helpful :/

File: src/lib/xpromoPersistState.js [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+    return checkDisplayStatus(statusCallback, isInterstitialDismissed);
+  });
+};
+
+/*
+ * Utils for sending callback once
+ * @TODO: this should be refactored,
+ * aspecialy const usedStatus usage...!
+ */ 
+const usedStatus = [];
```
Comment (schwers) [inline src/lib/xpromoPersistState.js]: 🔕  Would it be easier to make this an object, and look up the prescence of `status`? Then you could define `isStatusNotExist` as:

```js
const isStatusNotExist = status => !!usedStatus[status];
 ```

File: src/app/components/XPromoWrapper/index.jsx [hunk 19]
```diff
@@ -1,12 +1,80 @@
...
+  }
+
+  displayPersistBannerByTimer() {
+    const { dispatch, isInterstitialDismissed } = this.props;
+
+    xpromoPersist.runStatusCheck((status) => {
+      switch (status) {
+        case xpromoPersist.statusKey.JUST_DISMISSED: {
+          console.error('TESTING >', 'Event: cs.xpromo_view', ' -> ', 'just_dismissed_and_show');
+          dispatch(xpromoActions.trackXPromoEvent('cs.xpromo_view', {
```
Comment (schwers) [inline src/app/components/XPromoWrapper/index.jsx]: This could use the constant `XPROMO_VIEW` from `lib/eventUtils`

File: src/app/components/XPromoWrapper/index.jsx [hunk 20]
```diff
@@ -1,12 +1,80 @@
...
+            interstitial_type: 'persistent_banner',
+          }));
+          break;
+        }
+
+        case xpromoPersist.statusKey.SHOW: {
+          console.error('TESTING >', 'Event: cs.xpromo_view', ' -> ', 'show_banner');
+          dispatch(xpromoActions.promoPersistActivate());
+          dispatch(xpromoActions.trackXPromoEvent('cs.xpromo_view', {
```
Comment (schwers) [inline src/app/components/XPromoWrapper/index.jsx]: :nitpick: Please use the constant from `lib/eventUtils` here too

File: src/app/components/XPromoWrapper/index.jsx [hunk 21]
```diff
@@ -1,12 +1,80 @@
...
+            interstitial_type: 'persistent_banner',
+          }));
+          break;
+        }
+
+        case xpromoPersist.statusKey.NEW_SESSION: {
+          console.error('TESTING >', 'Event: cs.xpromo_ineligible', ' -> ', 'new_round_of_showing');
+          dispatch(xpromoActions.promoPersistActivate());
+          dispatch(xpromoActions.trackXPromoEvent('cs.xpromo_ineligible', {
```
Comment (schwers) [inline src/app/components/XPromoWrapper/index.jsx]: :nitpick: Please use the constant from lib/eventUtils here too

File: src/app/components/XPromoWrapper/index.jsx [hunk 22]
```diff
@@ -1,12 +1,80 @@
...
+            interstitial_type: 'persistent_banner',
+            inelligibility_reason: 'new_round_of_showing',
+          }));
+          break;
+        }
+
+        case xpromoPersist.statusKey.HIDE: {
+          console.error('TESTING >', 'Event: cs.xpromo_dismiss', ' -> ', 'hide_by_timeout');
+          dispatch(xpromoActions.promoPersistDeactivate());
+          dispatch(xpromoActions.trackXPromoEvent('cs.xpromo_dismiss', {
```
Comment (schwers) [inline src/app/components/XPromoWrapper/index.jsx]: :nitpick: Please use the constant from lib/eventUtils here too

File: src/app/components/XPromoWrapper/index.jsx [hunk 23]
```diff
@@ -1,12 +1,80 @@
...
+            interstitial_type: 'persistent_banner' ,
+            inelligibility_reason: 'hide_by_timeout',
+          }));
+          break;
+        }
+
+        default: {
+          console.error('TESTING >', 'Event: cs.xpromo_dismiss', ' -> ', 'recent_session');
+          dispatch(xpromoActions.promoPersistDeactivate());
+          dispatch(xpromoActions.trackXPromoEvent('cs.xpromo_dismiss', {
```
Comment (schwers) [inline src/app/components/XPromoWrapper/index.jsx]: :nitpick: Please use the constant from lib/eventUtils here too, last one :)

File: src/app/components/XPromoWrapper/index.jsx [hunk 24]
```diff
@@ -15,11 +83,11 @@ class XPromoWrapper extends React.Component {
...
       alreadyScrolledStart, 
       alreadyScrolledPast, 
-      xpromoThemeIsUsual, 
+      xpromoThemeIsUsual,
+      isXPromoPersistent,
     } = this.props;
 
-    const halfViewport = (window.pageYOffset > window.innerHeight / 2);
-
+    // @TODO dispatch(xpromoActions) to actions/xpomo.js
```
Comment (schwers) [inline src/app/components/XPromoWrapper/index.jsx]: 🔕  Is this todo still needed?

File: src/app/constants.js (hunk 12 above)
Comment (tomsoir) [inline src/app/constants.js]: Changed: 
xpromoDisplayTheme-> XPROMO_DISPLAY_THEMES
theme->COLOR_SCHEME

File: src/lib/timeChecker.js (hunk 17 above)
Comment (tomsoir) [inline src/lib/timeChecker.js]: Absolutely, 2-3 lines of code. All I'm worried about is the size of our final bundle file (~3,6Mb already). :(

**Comment by tomsoir** [comment]
💇‍♂️ @schwers

File: src/lib/xpromoState.js [hunk 25]
```diff
@@ -235,7 +241,6 @@ export const markListingClickTimestampLocalStorage = (dateTime) => {
   localStorage.setItem(XPROMO_LAST_MODAL_CLICK, dateTime);
 };
 
-
 function interstitialType(state) {
```
Comment (schwers) [inline src/lib/xpromoState.js]: Please modify this function to return 'persistent_banner' where its appropriate, and remove the hardcoding in `displayPersistBannerByTimer`

File: src/app/selectors/xpromo.js [hunk 26]
```diff
@@ -301,25 +319,32 @@ export function getXPromoExperimentPayload(state) {
   }
   return experimentPayload;
```
Comment (schwers) [inline src/app/selectors/xpromo.js]: This function needs to be updated to return the experiment data for the persistent experiment.

**Comment by tomsoir** [comment]
@schwers 💇‍♂️

File: src/app/components/XPromoWrapper/index.jsx [hunk 27]
```diff
@@ -34,46 +34,41 @@ class XPromoWrapper extends React.Component {
...
           break;
         }
 
         case xpromoPersist.statusKey.HIDE: {
           dispatch(xpromoActions.promoPersistDeactivate());
-          dispatch(xpromoActions.trackXPromoEvent(XPROMO_DISMISS, {
-            interstitial_type: 'persistent_banner' ,
-            ineligibility_reason: 'hide_by_timeout',
-          }));
+          dispatch(xpromoActions.trackXPromoEvent(
```
Comment (schwers) [inline src/app/components/XPromoWrapper/index.jsx]: 🔕 Due to the deactivate action, will this event and the ineligible event below have persistent experiment payload data?

Comment (tomsoir) [inline src/app/components/XPromoWrapper/index.jsx]: As I see, everything works well.

view_event
dismiss_event
ineligible_evevent

**Comment by tomsoir** [comment]
@schwers 💇‍♂️

File: src/app/featureFlags.js [hunk 28]
```diff
@@ -264,7 +264,10 @@ const config = {
       { allowedDevices: [IPHONE] },
       { allowedPages: ['index', 'listing', 'comments'] },
       { allowNSFW: false },
-      { not: { peak: 'mweb_xpromo_modal_listing_click_ios' } },
+      // @TODO: Add [] support ot { not: ...}
```
Comment (schwers) [inline src/app/featureFlags.js]: 🔕  (This is just a side-not, no effect on shipping this). I think we could get by with the `and` and `or` rules. e.g.

```js
{ not: { or: [
  { peak: 'mweb_xpromo_modal_listing_click_ios' },
  { peak: 'mweb_xpromo_require_login_ios' },
]}}
```

Comment (tomsoir) [inline src/app/featureFlags.js]: Oh, i will check right now. Thank you

Comment (tomsoir) [inline src/app/featureFlags.js]: It works well. Thank you
